File: swift/LoopflowCore/Models/TerminalSession.swift
import Foundation

public enum TerminalSessionStatus: String, Sendable, Codable, CaseIterable {
    case pending
    case attached
    case running
    case succeeded
    case failed
    case canceled

    public var isTerminal: Bool {
        switch self {
        case .succeeded, .failed, .canceled:
            true
        case .pending, .attached, .running:
            false
        }
    }
}

public struct TerminalSession: Sendable, Identifiable, Codable, Equatable {
    public let id: String
    public let waveId: String
    public let waveRunId: String?
    public let step: String
    public let agent: String
    public let cwd: String
    public let argv: [String]
    public let env: [String: String]
    public let source: String
    public let status: TerminalSessionStatus
    public let createdAt: Date
    public let attachedAt: Date?
    public let startedAt: Date?
    public let completedAt: Date?

    public init(
        id: String,
        waveId: String,
        waveRunId: String? = nil,
        step: String,
        agent: String,
        cwd: String,
        argv: [String] = [],
        env: [String: String] = [:],
        source: String = "wave_step",
        status: TerminalSessionStatus,
        createdAt: Date,
        attachedAt: Date? = nil,
        startedAt: Date? = nil,
        completedAt: Date? = nil
    ) {
        self.id = id
        self.waveId = waveId
        self.waveRunId = waveRunId
        self.step = step
        self.agent = agent
        self.cwd = cwd
        self.argv = argv
        self.env = env
        self.source = source
        self.status = status
        self.createdAt = createdAt
        self.attachedAt = attachedAt
        self.startedAt = startedAt
        self.completedAt = completedAt
    }
}

public struct TerminalLaunchSpec: Sendable, Equatable {
    public let sessionId: String
    public let waveId: String
    public let step: String
    public let agent: String
    public let cwd: String
    public let argv: [String]
    public let env: [String: String]
    public let completionToken: String

    public init(
        sessionId: String,
        waveId: String,
        step: String,
        agent: String,
        cwd: String,
        argv: [String],
        env: [String: String],
        completionToken: String
    ) {
        self.sessionId = sessionId
        self.waveId = waveId
        self.step = step
        self.agent = agent
        self.cwd = cwd
        self.argv = argv
        self.env = env
        self.completionToken = completionToken
    }
}

File: rust/loopflow/src/lfd/types/terminal_session.rs
use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};
use time::OffsetDateTime;

use crate::lfd::id::LfdId;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[non_exhaustive]
#[serde(rename_all = "snake_case")]
pub enum TerminalSessionStatus {
    Pending,
    Attached,
    Running,
    Succeeded,
    Failed,
    Canceled,
}

impl TerminalSessionStatus {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Pending => "pending",
            Self::Attached => "attached",
            Self::Running => "running",
            Self::Succeeded => "succeeded",
            Self::Failed => "failed",
            Self::Canceled => "canceled",
        }
    }

    pub(crate) fn as_i32(self) -> i32 {
        match self {
            Self::Pending => 0,
            Self::Attached => 1,
            Self::Running => 2,
            Self::Succeeded => 3,
            Self::Failed => 4,
            Self::Canceled => 5,
        }
    }

    pub(crate) fn from_i32(value: i32) -> Self {
        match value {
            0 => Self::Pending,
            1 => Self::Attached,
            2 => Self::Running,
            3 => Self::Succeeded,
            4 => Self::Failed,
            5 => Self::Canceled,
            _ => Self::Failed,
        }
    }

    pub fn from_exit_code(exit_code: i32) -> Self {
        if exit_code == 0 {
            Self::Succeeded
        } else {
            Self::Failed
        }
    }

    pub fn is_terminal(self) -> bool {
        matches!(self, Self::Succeeded | Self::Failed | Self::Canceled)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TerminalSession {
    pub id: LfdId,
    pub wave_id: LfdId,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub wave_run_id: Option<LfdId>,
    pub step: String,
    pub agent: String,
    pub cwd: String,
    #[serde(default)]
    pub argv: Vec<String>,
    #[serde(default, skip_serializing_if = "BTreeMap::is_empty")]
    pub env: BTreeMap<String, String>,
    pub source: String,
    pub status: TerminalSessionStatus,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub attached_at: Option<OffsetDateTime>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub started_at: Option<OffsetDateTime>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub completed_at: Option<OffsetDateTime>,
    pub created_at: OffsetDateTime,
    #[serde(skip_serializing)]
    pub completion_token: Option<String>,
}

impl TerminalSession {
    pub fn attach(&mut self) -> bool {
        if self.status != TerminalSessionStatus::Pending {
            return false;
        }
        self.status = TerminalSessionStatus::Attached;
        self.attached_at = Some(OffsetDateTime::now_utc());
        true
    }

    pub fn start(&mut self) -> bool {
        if self.status.is_terminal() {
            return false;
        }
        if self.attached_at.is_none() {
            self.attached_at = Some(OffsetDateTime::now_utc());
        }
        self.status = TerminalSessionStatus::Running;
        self.started_at = Some(OffsetDateTime::now_utc());
        true
    }

    pub fn complete(&mut self, exit_code: i32) -> bool {
        if self.status.is_terminal() {
            return false;
        }
        self.status = TerminalSessionStatus::from_exit_code(exit_code);
        self.completed_at = Some(OffsetDateTime::now_utc());
        true
    }

    pub fn cancel(&mut self) -> bool {
        if self.status.is_terminal() {
            return false;
        }
        self.status = TerminalSessionStatus::Canceled;
        self.completed_at = Some(OffsetDateTime::now_utc());
        true
    }
}

#[cfg(test)]
mod tests {
    use super::{TerminalSession, TerminalSessionStatus};
    use crate::lfd::id::LfdId;
    use time::OffsetDateTime;

    fn session(status: TerminalSessionStatus) -> TerminalSession {
        TerminalSession {
            id: LfdId::new(),
            wave_id: LfdId::new(),
            wave_run_id: None,
            step: "design".to_string(),
            agent: "claude".to_string(),
            cwd: "/tmp/repo".to_string(),
            argv: vec!["lf".to_string(), "design".to_string()],
            env: Default::default(),
            source: "wave_step".to_string(),
            status,
            attached_at: None,
            started_at: None,
            completed_at: None,
            created_at: OffsetDateTime::now_utc(),
            completion_token: None,
        }
    }

    #[test]
    fn attach_marks_pending_sessions_attached() {
        let mut session = session(TerminalSessionStatus::Pending);

        assert!(session.attach());
        assert_eq!(session.status, TerminalSessionStatus::Attached);
        assert!(session.attached_at.is_some());
    }

    #[test]
    fn start_auto_attaches_session() {
        let mut session = session(TerminalSessionStatus::Pending);

        assert!(session.start());
        assert_eq!(session.status, TerminalSessionStatus::Running);
        assert!(session.attached_at.is_some());
        assert!(session.started_at.is_some());
    }

    #[test]
    fn terminal_sessions_do_not_restart_or_complete_twice() {
        let mut session = session(TerminalSessionStatus::Succeeded);

        assert!(!session.start());
        assert!(!session.complete(1));
        assert!(!session.cancel());
    }
}
